BLE: Security Manager traces

pan- opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Traces added to Bluetooth Security Manager. 

- [x] Complete SecurityManager
- [x] Move String serialization into own file
- [x] Add traces to PAL

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->



### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- committed:
WIP: SM Bluetooth traces

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -77,8 +189,17 @@ ble_error_t SecurityManager::init(
+        to_string(iocaps),
+        passkey_str(passkey),
+        to_string(signing),
+        db_path
```
~it's a string, though I understand it can be null~ nevermind it's traced later

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -205,7 +335,11 @@ ble_error_t SecurityManager::preserveBondingStateOnReset(bool enabled)
+    if (!_db) {
+        tr_error("Failure, DB not initialized");
+        return BLE_ERROR_INITIALIZATION_INCOMPLETE;
+    }
```
line 349 can fail without trace if _pal.set_csrk doesn't log traces

pan- replied:
But it will log trace, generaly, when the step is `error = operation; if (error) { return error; }` We can assume (in this context) that operation will trace the error and details of the call. 

I don't want to trace errors twice if it doesn't add meaningful information.

paul-szczepanek-arm replied:
yeah, as long as tracing is also added to pal

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -363,44 +518,70 @@ ble_error_t SecurityManager::acceptPairingRequest(connection_handle_t connection
+                    identity ? identity->identity_address_is_public : false
+                );
+            } else {
+                tr_warning("Event handler removing while retrieving peer ID address");
```
```suggestion
                tr_warning("Event handler removed while retrieving peer ID address");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -468,13 +662,21 @@ ble_error_t SecurityManager::setLinkSecurity(
     }
 
     if (cb->encryption_requested) {
+        tr_error("Failure, encryption already requested");
```
wouldn't class it as an error, since it's already pending you will actually get what you wnted

paul-szczepanek-arm replied:
actually no, you could ask for something more

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -537,18 +750,21 @@ ble_error_t SecurityManager::enableSigning(
                 cb->db_entry
             );
         } else {
+            tr_info("CSRK not available, pair to acquire it");
```
```suggestion
            tr_info("CSRK not available, pairing to acquire it");
```
otherwise it implies user action was needed

paul-szczepanek-arm replied:
actually if you're slave you can't force that so maybe the trace should be in the if statement where the role is known

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -577,14 +794,19 @@ ble_error_t SecurityManager::getLinkEncryption(
 
     SecurityDistributionFlags_t* flags = _db->get_distribution_flags(cb->db_entry);
     if (!flags) {
+        tr_error("Failed to retrieve link encryption, empty distribution flags for %d", connection);
```
now I think about it, this can't ever happen, if entry exists it contains flags

pan- replied:
It is present in many places, it can happen if the database is corrupted. I wonder if we shouldn't check it when it is loaded then assume it is consistent.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -613,30 +835,38 @@ ble_error_t SecurityManager::setLinkEncryption(
 )
 {
-    if (!_db) return BLE_ERROR_INITIALIZATION_INCOMPLETE;
+    tr_info("Set %d encryption to %s", connection, to_string(encryption));
```
The problem with logs that are only meaningful in pairs is that tracing may be broken up by other traces and becomes hard to read when the meaning of a line depends on a line potentially N lines lower and you will not know if it's there or which statement it links to.

I think each trace should be self contained.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -672,10 +904,13 @@ ble_error_t SecurityManager::setLinkEncryption(
     }
 
-    eventHandler->linkEncryptionResult(connection, current_encryption);
+    if (eventHandler) {
```
nice catch, maybe worth a separate commit

paul-szczepanek-arm reviewed (approved):
Very comprehensive. I have my doubts about the context sensitive traces approach.

I don't like tracing whose interpretations relies on more than one line, lines can be broken up and then you can't tell if an action completed correctly or not because even if there is a error trace then you might connect it to the wrong trace line.

Exception are OK, like for example it doesn't matter what causes the "DB not initialised" trace, because it's a global db so it just means it's not there, doesn't matter what caused the discovery of it being missing.

pan- committed:
BLE: Move traces helper into their own file

pan- committed:
BLE: Move traces helper into their own file

pan- committed:
BLE: Add traces to Security Manager PAL

pan- committed:
BLE: Make SM traces consistent.
- Prefix with Connection <id> - when appropriate.
- Display parameters after `:`.
- If multiple parameters should be displayed name then and print the value after =. They are separated by a `,`.

pan- committed:
BLE: Fix SM random number generation.

Only the first 8 bytes were generated from the stack.

pan- committed:
BLE: Regenerate CSRK if it is all zeroes.
